lib/Mojo/Message.pm

```diff
@@ -132,7 +132,8 @@ sub is_limit_exceeded { !!shift->{limit} }
 sub json {
   my ($self, $pointer) = @_;
   return undef if $self->content->is_multipart;
-  my $data = $self->{json} //= j($self->body);
+  my $charset = $self->body_params->charset || $self->default_charset;
+  my $data    = $self->{json} //= j($charset eq 'UTF-8' ? $self->body : encode('UTF-8', decode($charset, $self->body)));
```

Comment:
`j(encode(decode()))` seems super inefficient.

Reply:
Agreed. Tried this before:
```perl
my $data = $charset eq 'UTF-8' ? j($self->body) : from_json(decode($charset,$self->body));
```
which seems more natural but breaks the last test in t/mojo/response.t subtest "Parse response and extract JSON data"
```
t/mojo/response.t .......................... 1/?
    #   Failed test 'right result'
    #   at t/mojo/response.t line 1075.
    #     Structures begin differing at:
    #          $got->[1] = '2'
    #     $expected->[1] = '4'
    # Looks like you failed 1 test of 9.

#   Failed test 'Parse response and extract JSON data'
#   at t/mojo/response.t line 1076.
# Looks like you failed 1 test of 56.
```
i.e. this one:
```perl
  is_deeply $res->json('/baz'), [1, 2, 3], 'right result';
  $res->json->{baz}[1] = 4;
  is_deeply $res->json('/baz'), [1, 4, 3], 'right result';
```

`j(encode(decode()))` is awkward but avoids that subtle breakage.